### PR TITLE
Offset Implementation for Text and Block

### PR DESCRIPTION
**Pull request by Josef-Vonasek**
### Pull Request Description

Computes actual offsets for text literals, blocks and modules.

### Checklist
Please include the following checklist in your PR:

- [x] The documentation has been updated if necessary.
- [x] All code conforms to the [Scala](https://github.com/luna/enso/blob/master/doc/scala-style-guide.md), [Java](https://github.com/luna/enso/blob/master/doc/java-style-guide.md) or [Haskell](https://github.com/luna/enso/blob/master/doc/haskell-style-guide.md) style guides as appropriate.
- [x] All code has been tested where possible.

### CODE REVIEW

**Comment by iamrecursion** [comment]
Please label your PR. I'll review it soon.

File: Syntax/definition/src/main/scala/org/enso/data/Implicits.scala [hunk 1]
```diff
@@ -0,0 +1,16 @@
+package org.enso.data
+
+object Implicits {
+
+  implicit def just[A](a: A): Just[A] = new Just(a)
+
+  class Just[A](val a: A) extends AnyVal {
+
+    /** Returns value, then performs an effect.
```
Comment (iamrecursion) [inline Syntax/definition/src/main/scala/org/enso/data/Implicits.scala]:
```suggestion
    /** Returns a value then performs an effect.
```

File: Syntax/definition/src/main/scala/org/enso/data/Implicits.scala [hunk 2]
```diff
@@ -0,0 +1,16 @@
...
+      *
+      * Usage: value.thenDo { .. effect .. }
+      */
+    def thenDo(effect: => Unit): A = { effect; a }
```
Comment (iamrecursion) [inline Syntax/definition/src/main/scala/org/enso/data/Implicits.scala]: I'm stunned this function doesn't exist somewhere already.

Comment (Josef-Vonasek) [reply Syntax/definition/src/main/scala/org/enso/data/Implicits.scala]: Me too!

File: Syntax/definition/src/main/scala/org/enso/data/Implicits.scala [hunk 3]
```diff
@@ -0,0 +1,14 @@
+package org.enso.data
+
+object Implicits {
+
+  implicit class Just[A](val a: A) extends AnyVal {
```
Comment (wdanilo) [inline Syntax/definition/src/main/scala/org/enso/data/Implicits.scala]: Why this class is named "Just"? It makes it easy for the reader to think it is the extension for "Just", not for every type.

File: Syntax/definition/src/main/scala/org/enso/syntax/text/AST.scala [hunk 4]
```diff
@@ -1026,15 +1030,26 @@ object AST {
...
+        t.copy(
+          firstLine = t.firstLine.copy(elem = (0, t.firstLine.elem)),
+          lines = {
+            var offset = t.firstLine.span
+            for (line <- t.lines)
+              yield line.copy(elem = line.elem.map((offset, _))).thenDo {
+                offset += line.span
+              }
+          }
+        )
```
Comment (wdanilo) [inline Syntax/definition/src/main/scala/org/enso/syntax/text/AST.scala]: TBH I jsut hate this formatting style (caused by how the code is written). I owuld never write such long expression by myself. Imo such code (and all above codes) would be a much better written this way:

```scala
val firstLine2 = t.firstLine.copy(elem = (0, t.firstLine.elem))
var offset     = t.firstLine.span
val lines2     = for (line <- t.lines) {
    val elem2 = line.elem.map((offset, _))
    val line2 = line.copy(elem = elem2)
    offset += line.span
    yield line2
}
t.copy(firstLine = firstLine2, lines = lines2)
```
This way you separate chunks fo the code to separate variables, its much easier to track whats happening in the code, it has smaller indentation and uses the same amount of lines.

Comment (Josef-Vonasek) [reply Syntax/definition/src/main/scala/org/enso/syntax/text/AST.scala]: Ugly indeed. I will rewrite it.

**Comment by iamrecursion** [comment]
Merging master in to rerun CI with the new config.

**Pull request merged**